### DISCUSSION TITLE
Can someone explain how select/poll/epoll interacts with C# code in socket Select() or SocketAsyncEventArgs?

### DISCUSSION THREAD

**Discussion by cjmld5** [dotnet/runtime#66932] [Q&A]
Can someone explain how select/poll/epoll interacts with C# code in socket Select() or SocketAsyncEventArgs?
below example:
```
Socket.Select() 
-> UnsafeNclNativeMethods.OSSOCK.select(0, array, array2, array3, ref socketTime)
-> [DllImport("ws2_32.dll", SetLastError = true)]
     internal static extern int select([In] int ignoredParameter, [In][Out] IntPtr[] readfds, [In][Out] IntPtr[] writefds, [In][Out] IntPtr[] exceptfds, [In] ref TimeValue timeout)
```
and how winsock2 interacts with dotnet runtime?it seems in C++ we could add a head
`#include<winsock2.h>`
and see select code below:
```
#if INCL_WINSOCK_API_PROTOTYPES
WINSOCK_API_LINKAGE
int
WSAAPI
select(
    _In_ int nfds,
    _Inout_opt_ fd_set FAR * readfds,
    _Inout_opt_ fd_set FAR * writefds,
    _Inout_opt_ fd_set FAR * exceptfds,
    _In_opt_ const struct timeval FAR * timeout
    );
#endif /* INCL_WINSOCK_API_PROTOTYPES */
```
and here could not find some relations to interacts with dotnet runtime pal_networking.c？How should they be strung together？

**Comment by huoyaoyuan** [comment]
@dotnet/ncl

**Comment by wfurt** [comment]
That really depends on how the socket is used from c# @cjmld5. For example, you can use synchronous read and write and neither one would be used. You can use `Socket.Select` or `Socket.Poll` and that would translate pretty much directly to OS calls. The `epoll` comes in picture when Async methods are used. In ideal case, I would suggest not to mix sync & async - pick one and stick to it.

**Reply by cjmld5** [reply]
yes，this should not be a better practice that mix usage of sync & async. it seems that it is a knowledge blind spot for me on how  Socket.Select or Socket.Poll translate to OS calls：**Socket.Select** gose to **winsock2** and how **winsock2** go to **ws2_32.dll**`s select method, and how **winsock2** goes to https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/Interop/Windows/WinSock/Interop.select.cs  or other platform *.cs

**Comment by ManickaP** [comment]
About the `select` and diff between *NIX and Win: on Windows we p/invoke directly into system libraries without the native shims (pal_networking.c): https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/Interop/Windows/WinSock/Interop.select.cs since Windows has stable APIs.

I hope this helps a bit.

**Reply by cjmld5** [reply]
could u explain more about "**we p/invoke directly into system libraries without the native shims**" like how `select `Eventually goes to "Interop.select.cs"？
